lib/semantics/tools.cc

```diff
@@ -536,6 +537,150 @@ std::unique_ptr<parser::Message> WhyNotModifiable(parser::CharBlock at,
   return {};
 }
 
+static bool IsCoarray(const parser::AllocateObject &allocateObject) {
+  const parser::Name &name{GetLastName(allocateObject)};
+  return name.symbol && IsCoarray(*name.symbol);
+}
+
+static bool IsImageControlStmt(const parser::AllocateStmt &allocateStmt) {
+  const auto &allocationList{
+      std::get<std::list<parser::Allocation>>(allocateStmt.t)};
+  for (const auto &allocation : allocationList) {
+    const auto &allocateObject{std::get<parser::AllocateObject>(allocation.t)};
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool IsImageControlStmt(const parser::DeallocateStmt &deallocateStmt) {
+  const auto &allocateObjectList{
+      std::get<std::list<parser::AllocateObject>>(deallocateStmt.t)};
+  for (const auto &allocateObject : allocateObjectList) {
+    if (IsCoarray(allocateObject)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static bool IsImageControlStmt(const parser::CallStmt &callStmt) {
+  const auto &procedureDesignator{
+      std::get<parser::ProcedureDesignator>(callStmt.v.t)};
+  if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+    // TODO: also ensure that the procedure is, in fact, an intrinsic
+    if (name->source == "move_alloc") {
+      const auto &args{
+          std::get<std::list<parser::ActualArgSpec>>(callStmt.v.t)};
+      if (!args.empty()) {
+        const parser::ActualArg &actualArg{
+            std::get<parser::ActualArg>(args.front().t)};
+        if (const auto *argExpr{
+                std::get_if<common::Indirection<parser::Expr>>(&actualArg.u)}) {
+          return HasCoarray(argExpr->value());
+        }
+      }
+    }
+  }
+  return false;
+}
+
+struct ImageControlStmtHelper {
+  using ImageControlStmts = std::variant<parser::ChangeTeamConstruct,
+      parser::CriticalConstruct, parser::EventPostStmt, parser::EventWaitStmt,
+      parser::FormTeamStmt, parser::LockStmt, parser::StopStmt,
+      parser::SyncAllStmt, parser::SyncImagesStmt, parser::SyncMemoryStmt,
+      parser::SyncTeamStmt, parser::UnlockStmt>;
+  template<typename T> bool operator()(const T &) {
+    return common::HasMember<T, ImageControlStmts>;
+  }
+  template<typename T> bool operator()(const common::Indirection<T> &x) {
+    return (*this)(x.value());
+  }
+  bool operator()(const parser::AllocateStmt &stmt) {
+    return IsImageControlStmt(stmt);
+  }
+  bool operator()(const parser::DeallocateStmt &stmt) {
+    return IsImageControlStmt(stmt);
+  }
+  bool operator()(const parser::CallStmt &stmt) {
+    return IsImageControlStmt(stmt);
+  }
+  bool operator()(const parser::Statement<parser::ActionStmt> &stmt) {
+    return std::visit(*this, stmt.statement.u);
+  }
+};
+
+bool IsImageControlStmt(const parser::ExecutableConstruct &construct) {
+  return std::visit(ImageControlStmtHelper{}, construct.u);
+}
+
+using ResultType = std::optional<parser::MessageFixedText>;
```

Comment:
This would be okay within the scope of `GetImageControlStmtCoarrayMsg` but I don't think it is appropriate at the file level. It's the result type of only one of the many functions declared here.

Reply:
Sounds good.  I'll make that change.